tests/decap/test_decap.py

```diff
@@ -0,0 +1,242 @@
+        ip = IPNetwork(addr).ip
+        if ip.version == 4 and not lo_ip:
+            lo_ip = ip
+        elif ip.version == 6 and not lo_ipv6:
+            lo_ipv6 = ip
+    logger.info("lo_ip={}, lo_ipv6={}".format(str(lo_ip), str(lo_ipv6)))
+
+    vlan_ip = None
+    vlan_ipv6 = None
+    if "VLAN_INTERFACE" in cfg_facts:
+        for addr in cfg_facts["VLAN_INTERFACE"]["Vlan1000"]:
+            ip = IPNetwork(addr).ip
+            if ip.version == 4 and not vlan_ip:
+                vlan_ip = ip
+            elif ip.version == 6 and not vlan_ipv6:
+                vlan_ipv6 = ip
+    logger.info("vlan_ip={}, vlan_ipv6={}".format(str(vlan_ip), str(vlan_ipv6)))
+
+    # config decap
+    decap_conf_template = Template(open("../ansible/roles/test/templates/decap_conf.j2").read())
+
+    src_ports = set()
+    topology = testbed["topo"]["properties"]["topology"]
+    if "host_interfaces" in topology:
+        src_ports.update(topology["host_interfaces"])
+    if "disabled_host_interfaces" in topology:
+        for intf in topology["disabled_host_interfaces"]:
+            src_ports.discard(intf)
+    if "VMs" in topology:
+        for k, v in topology["VMs"].items():
+            src_ports.update(v["vlans"])
+
+    decap_conf_vars = {
+        "outer_ipv4": to_bool(request.config.getoption("outer_ipv4")),
+        "outer_ipv6": to_bool(request.config.getoption("outer_ipv6")),
+        "inner_ipv4": to_bool(request.config.getoption("inner_ipv4")),
+        "inner_ipv6": to_bool(request.config.getoption("inner_ipv6")),
+        "lo_ip": str(lo_ip),
+        "lo_ipv6": str(lo_ipv6),
+        "op": "SET",
+        "dscp_mode": dscp_mode,
+        "ecn_mode": ecn_mode,
+        "ttl_mode": ttl_mode,
+    }
+
+    duthost.copy(content=decap_conf_template.render(**decap_conf_vars), dest="/tmp/decap_conf.json")
+    duthost.shell("docker cp /tmp/decap_conf.json swss:/decap_conf.json")
+    duthost.shell('docker exec swss sh -c "swssconfig /decap_conf.json"')
+
+    # Prepare PTFf docker
+    prepare_ptf(ptfhost, testbed, cfg_facts)
+
+    setup_info = {
+        "src_ports": ",".join([str(port) for port in src_ports]),
+        "router_mac": cfg_facts["DEVICE_METADATA"]["localhost"]["mac"],
+        "vlan_ip": str(vlan_ip) if vlan_ip else "",
+        "vlan_ipv6": str(vlan_ipv6) if vlan_ipv6 else "",
+    }
+    setup_info.update(decap_conf_vars)
+    logger.info(json.dumps(setup_info, indent=2))
+
+    yield setup_info
+
+    # Remove decap configuration
+    decap_conf_vars["op"] = "DEL"
+    duthost.copy(content=decap_conf_template.render(**decap_conf_vars), dest="/tmp/decap_conf.json")
+    duthost.shell("docker cp /tmp/decap_conf.json swss:/decap_conf.json")
+    duthost.shell('docker exec swss sh -c "swssconfig /decap_conf.json"')
+
+
+def test_decap(setup_teardown, testbed, ptfhost):
+
+    setup_info = setup_teardown
+
+    log_file = "/tmp/decap.{}.log".format(datetime.now().strftime('%Y-%m-%d-%H:%M:%S'))
+    ptf_runner(ptfhost,
```

Comment:
What do you think about removing ptfrunner from pytest? I think it was introduced to run ptf from within Ansible and it is not longer required when we can call ptfhost.shell. 

I took first step here: https://github.com/Azure/sonic-mgmt/blob/master/tests/common/fixtures/advanced_reboot.py#L125

Reply:
Under the hood, the ptf_runner uses ptfhost.shell to run the ptf script. I think the ptf_runner makes the building of ptf script cli easier and we can keep it.
BTW, I don't understand how the example you given is related with removing ptf_runner. Could you please elaborate?